### PR TITLE
Validate MIDI range, (fixes  #69)

### LINKED ISSUE #69: Validate MIDI note range
Currently, attempting to play a note that is out of the MIDI note range (0-127) results in a confusing error message:

```
$ alda play -c 'piano: o10 c'
[27713] Parsing/evaluating...
[27713] ERROR data1 out of range: 132
```

We should catch this and throw our own error with a clearer message.

### PR DESCRIPTION
**Pull request by iggar**
(This fixes #69)
Validate MIDI range, raise exception when out of range. Replace 'if' with more idiomatic 'when'. Tidy up code. Add note_test.
Note: on `alda.lisp.events.note-test` I've managed to add the positive case test ("happy path") but I couldn't figure out the negative test case, when the exception is thrown. So I left it commented out for discussion. 
I'm happy to make further changes, e.g. the text on the message to be shown

### CODE REVIEW

File: src/alda/lisp/events/note.clj [hunk 1]
```diff
@@ -37,22 +37,30 @@
...
+          midi-note-val      (when note-event?
                                (determine-midi-note event octave key-signature
                                                     transposition))
-          pitch              (if (= event-type :note)
+          midi-note          (if (or (nil? midi-note-val) (<= 0 midi-note-val 127))
+                               midi-note-val
+                               (throw
+                                (ex-info (str "The calculated MIDI note "
+                                              "(" midi-note-val ") is out "
+                                              "of the allowed range.")
```
Comment (daveyarwood) [inline src/alda/lisp/events/note.clj:49]:
```suggestion
                                              "of the allowed range (0-127).")
```

File: src/alda/lisp/events/note.clj [hunk 2]
```diff
@@ -37,22 +37,30 @@
...
+          note-event?        (= event-type :note)
+          midi-note-val      (when note-event?
                                (determine-midi-note event octave key-signature
                                                     transposition))
-          pitch              (if (= event-type :note)
+          midi-note          (if (or (nil? midi-note-val) (<= 0 midi-note-val 127))
+                               midi-note-val
+                               (throw
+                                (ex-info (str "The calculated MIDI note "
+                                              "(" midi-note-val ") is out "
```
Comment (daveyarwood) [inline src/alda/lisp/events/note.clj:48]:
```suggestion
                                              "(" midi-note-val ") is outside "
```

File: src/alda/lisp/events/note.clj [hunk 3]
```diff
@@ -37,22 +37,30 @@
...
                                                  ms)
           quant-duration     (* full-duration quant)
-          midi-note          (if (= event-type :note)
+          note-event?        (= event-type :note)
+          midi-note-val      (when note-event?
                                (determine-midi-note event octave key-signature
                                                     transposition))
-          pitch              (if (= event-type :note)
+          midi-note          (if (or (nil? midi-note-val) (<= 0 midi-note-val 127))
+                               midi-note-val
```
Comment (daveyarwood) [inline src/alda/lisp/events/note.clj:45]: Essentially, what we're doing here is setting up a binding `midi-note`, which will either end up being the same value as `midi-note-val` or it will throw an error.

I think we could slightly simplify this code by calling the calculation on lines 41-42 `midi-note` instead of `midi-note-val`, and then inserting a let-binding pair where we just throw an error if appropriate:

```clojure
midi-note (when note-event?
            (determine-midi-note ...)
_         (when (and midi-note (not (<= 0 midi-note 127)))
            (throw ...))
```

File: test/alda/lisp/events/note_test.clj [hunk 4]
```diff
@@ -0,0 +1,78 @@
+(ns alda.lisp.events.note-test
+  (:require [alda.lisp.events.note :as note]
+            [clojure.test :refer :all])
+  (:import [alda.lisp.model.records AbsoluteOffset]))
+
+(deftest event-updates-test
+  (testing "MIDI note inside range parsing correctly"
+    (let [score {:chord-mode false
```
Comment (daveyarwood) [inline test/alda/lisp/events/note_test.clj:8]: I worry that representing the value of the score map inline here might be coupling too closely to the current implementation. I think this should be expressed instead of in terms of behavior/actions.

In this case, I think we just want to initialize a score and add a piano to it, so we can represent that as `(score (part "piano"))`.

File: test/alda/lisp/events/note_test.clj [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+                                :time-scaling 1
+                                :stock "midi-acoustic-grand-piano"
+                                :track-volume 0.7874015748031497}}
+                 :markers {:start 0}
+                 :cram-level 0
+                 :global-attributes {}
+                 :tempo/values {0 120}
+                 :nicknames {}
+                 :beats-tally-default nil}
+          event {:event-type :note :letter :c :accidentals []
```
Comment (daveyarwood) [inline test/alda/lisp/events/note_test.clj:37]: Similarly, we can (and should, in my opinion) represent this as `(note (pitch :c))` instead of inlining the map that that function call returns.

File: test/alda/lisp/events/note_test.clj [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+                                  :track-volume 0.7874015748031497}}
+                   :markers {:start 0}
+                   :cram-level 0
+                   :global-attributes {}
+                   :tempo/values {0 120}
+                   :nicknames {}
+                   :beats-tally-default nil}
+            event {:event-type :note :letter :c :accidentals []
+                   :beats nil :ms nil :slur? nil}]
+        (is (thrown? Exception (note/event-updates score event))))))
```
Comment (daveyarwood) [inline test/alda/lisp/events/note_test.clj:78]: This seems to me like a totally reasonable way to test the sad path where the exception is thrown. A couple notes:

* The note needs to be outside of the acceptable MIDI range in order to produce the exception, so one way to achieve this is to start with a new score, then add an "octave set" event to set the octave to a really high or really low octave, then add a C note.

  Here are a couple examples of how to do this in an Alda REPL:

  ```
  > piano: (octave 10) c
  [27713] ERROR data1 out of range: 132
  > piano: (octave -2) c
  [27713] ERROR data1 out of range: -12
  ```

  I think the corresponding Clojure code for these tests would be something like:

  ```clojure
  (let [score (score (part "piano") (octave -2))
        event (note (pitch :c))]
    (is (thrown? Exception (note/event-updates score event))
  ```

Comment (daveyarwood) [reply test/alda/lisp/events/note_test.clj:78]: * Another thing worth considering is whether it might make sense to test that the exception we have in mind is actually being thrown. I've used `thrown-with-msg` for this in the past. In this case, we could maybe do something like:

  ```clojure
  (is (thrown-with-msg?
        clojure.lang.ExceptionInfo
        #"outside of the allowed range"
        (note/event-updates score event)))
  ```

**Comment by daveyarwood** [comment]
Hey @iggar , thanks for the PR! This looks like it should do the trick. Nice work so far. I've left some comments/suggestions.

**Comment by daveyarwood** [comment]
At this point in the saga of Alda, the focus has shifted to a new [v2 branch](https://github.com/alda-lang/alda/tree/v2) implemented in Go and Kotlin. Alda v2 will be released soon (:crossed_fingers: ) so I think we can close the PR at this point. I am keeping the associated issue open for now until I can confirm the same issue is fixed in Alda v2.
